superset-frontend/src/dashboard/util/crossFilters.ts

```diff
@@ -52,6 +52,20 @@ export const getCrossFiltersConfiguration = (
     return undefined;
   }
 
+  const chartsByDataSource: Record<string, number[]> = Object.values(
+    charts,
+  ).reduce((acc: Record<string, number[]>, chart) => {
+    if (!chart.form_data) {
+      return acc;
+    }
+    const { datasource } = chart.form_data;
+    if (!acc[datasource]) {
+      acc[datasource] = [];
+    }
+    acc[datasource].push(chart.id);
```

Comment:
Using a set would likely make the algorithm more efficient.
```suggestion
    if (!acc[datasource]) {
      acc[datasource] = new Set();
    }
    acc[datasource].add(chart.id);
```